feat(translation): Init markdown translation.

IRONICBo opened the pull request:
Add translation pkg to goplus community.

Ref issue: https://github.com/goplus/community/issues/31

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,60 @@
+	pointer := doc.FirstChild()
+	for pointer != nil {
+		pointer = pointer.NextSibling()
+		raw_text := extractText(pointer, []byte(src))
```
use `rawText`

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+// Translate translate sequence of text
+func TranslateSeq(src string, from, to Language) (string, error) {
```
maybe should support batch translate to reduce rpc-call times.

IRONICBo replied:
Thanks! I might consider using a map-reduce model approach to speed up the translation process.

IRONICBo replied:
Fix :)

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,179 @@
+// TranslateConfig is the config of translation server
+type TranslateConfig struct {
+	apiKey string // api key of translation server
+}
```
```go
type Config struct {
    APIKey string
}
```

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+// NewTranslateConfig create a new TranslateConfig
+func NewTranslateConfig(apiKey string) *TranslateConfig {
```
```go
type Engine struct {
    // ...
}

func New(config *Config) *Engine
```

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+// TranslateMarkdown translate markdown text
+func (c *TranslateConfig) TranslateMarkdown(src string, from, to Language) (string, error) {
```
Can provide both `[]byte` and `string` version.

```go
func (p *Engine) TranslateMarkdown(src []byte, from, to Language) (ret []byte, err error)
func (p *Engine) TranslateMarkdownText(src string, from, to Language) (ret string, err error)
```

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+// Translate translate sequence of text
+func (c *TranslateConfig) TranslateSeq(src string, from, to Language) (string, error) {
```
```go
func (p *Engine) TranslatePlain(src []byte, from, to Language) (ret []byte, err error)
func (p *Engine) TranslatePlainText(src string, from, to Language) (ret string, err error)
```

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+// TranslateBatchSeq translate a series of sequences of text
+func (c *TranslateConfig) TranslateBatchSeq(src []string, from, to Language) ([]string, error) {
```
后来想到一个方法可能可以避免需要显式对 batch 能力的依赖。
比如按如下步骤来做：
* 对 (markdown []byte) 进行分割，得到 (ops [][]byte, vals [][]byte)。这里 markdown = `ops[0] vals[0] ops[1] vals[1] ... ops[n-1] vals[n-1]`，其中 ops 数组不需要翻译，但是可能分 3 种语义：1）为 nil 或 markdown 控制字符，此时 val 为普通文本。2）特殊值 opEscape，表示 val 在 markdown 原文是转义的，这里存它 decode 后得到的文本。3）我们不希望出现的特殊序列。例如 markdown 中连续出现的 n 个回车 "\n"。那么 op 为 n-1 个 "\n"，而对应的 val 为单个回车 "\n"。
* 对 (vals [][]byte) 进行用特定在 vals 中绝对不会出现的分隔符 (sep []byte) 连接：text := bytes.Join(vals, sep)。比如我们定义 sep = "\n\n"。
* 对 text 整体进行翻译，得到 result := translate(text, from, to)。
* 对 result 按 sep 进行 split，得到 rets := bytes.Split(result, sep)。确认 len(rets) == len(ops) == len(vals)。选择 sep 最大的讲究是要确保 translate 不会翻译 sep，会被原样保留在翻译后的文本中。
* 将 (ops [][]byte, rets [][]byte) 重新组装回 markdown，结束。

xushiwei replied:
```
3）我们不希望出现的特殊序列。例如 markdown 中连续出现的 n 个回车 "\n"。那么 op 为 n-1 个 "\n"，而对应的 val 为单个回车 "\n"。
```
这一段有一个需要注意的细节：要避免多个 val 连在一起后包含 sep。所以这里更好的做法是：
```
3）我们不希望出现的特殊序列。例如 markdown 中连续出现的 n 个回车 "\n"。那么 op 为 n 个 "\n"，而对应的 val 为 nil。
```

IRONICBo replied:
感谢您的建议：
我使用goldmark中可以获取到Text类型在原文件中的索引，所以我将opt换成了对应文本的索引位置。同时，使用随机生成的分隔符来拆分不同文档的索引位置。

目前对翻译前后的数组控制的比较严格，后期可能对不同翻译位置做更细粒度的控制，减少匹配失败和翻译重试。

---

I used goldmark in which I could get the index of the Text type in the original file, so I replaced opt with the index position of the corresponding text. Also, a randomly generated separator is used to split the index positions of the different documents.

At present, the control of the array before and after translation is relatively strict, later may do more granular control of different translation positions to reduce the matching failure and translation retry.

IRONICBo commented:
TestTranslateMarkdown test result.

<pre>
# Heading 


## This is a secondary title

This is an ordinary text.

This is a paragraph*Bold font*Gets or sets the text of the.

This is a paragraph**Bold font**Gets or sets the text of the.

This is a paragraph***Rough italics***Gets or sets the text of the.

This is a paragraph`Inline code`.

This is a block of code:

```go
func main() {
        fmt.Println("Hello, World!")
}
```

Here's a list:

- List Item 1
- List Item 2
        - List item 2.1
        - Table item 2.2
- List Item 3

This is an ordered list:

1. List Item 1
2. List Item 2
        1. List item 2.1
        2. Table item 2.2
3. List Item 3

This is a reference:

> This is a quote.

Here is an indented code block:

        func main() {
                fmt.Println("Hello, World!")
        }

This is a paragraph[Link](https://www.example.com)

This is a paragraph![Image](https://www.example.com/image.jpg)
</pre>

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,217 @@
+
+// Engine is the config of translation server
+type Engine struct {
+	APIKey string // api key of translation server
```
don't need public this member

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+// TranslateResponse is the response of translation server
+type TranslateResponse struct {
```
don't public this type

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,217 @@
+
+// TranslateBatchSeq translate a series of sequences of text
+// Deprecated: use TranslatePlainXXX instead
+func (c *Engine) TranslateBatchSeq(src []string, from, to language.Tag) ([]string, error) {
```
remove this func

IRONICBo replied:
Translation API has 5000 characters limit, so I need to split the text to batch plains.

I will update it to support long markdown text translation. @xushiwei

xushiwei replied:
batch can be a internal mechanism. `TranslatePlain` should have no characters limit.

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,217 @@
+	var req *http.Request
+	req, err = http.NewRequest("POST", niuTransAPI, strings.NewReader(formData.Encode()))
+	if err != nil {
+		return "", ErrRequestFailed
```
don't shadow detail error

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,217 @@
+
+	if resp, err := http.DefaultClient.Do(req); err != nil {
+		return "", ErrRequestFailed
+	} else {
```
don't use `else`

xushiwei commented on translation/markdown.go:
```diff
@@ -0,0 +1,217 @@
+	toBeTranslatedStr := strings.Join(translationVec, fmt.Sprintf("\n%s\n", translationSep))
+	translatedStr, err := c.TranslatePlainText(toBeTranslatedStr, from, to)
+	if err != nil {
+		return []byte(""), err
```
return `nil` instead of `[]byte("")`

IRONICBo commented:
- Support long markdown translation(Over 5000 characters)

Translation result:
<pre>
=== RUN   TestTranslateMarkdown

# Heading 


## This is a secondary title

This is an ordinary text.

This is a paragraph*Bold font*Gets or sets the text of the.

This is a paragraph**Bold font**Gets or sets the text of the.

This is a paragraph***Rough italics***Gets or sets the text of the.

This is a paragraph`Inline code`.

This is a block of code:

```go
func main() {
	fmt.Println("Hello, World!")
}
```

Here's a list:

- List Item 1
- List Item 2
	- List item 2.1
	- Table item 2.2
- List Item 3

This is an ordered list:

1. List Item 1
2. List Item 2
	1. List item 2.1
	2. Table item 2.2
3. List Item 3

This is a reference:

> This is a quote.

Here is an indented code block:

	func main() {
		fmt.Println("Hello, World!")
	}

Xu Shiwei: Design and implementation of Go + v1. x, the first phase of Go + open class
Original Seven Niuyun Technical Team Seven Niuyun Technical Team 2021-11-12 17:03
On October 15th, Qi Niuyun hosted "Go + Together! Go + 1.0 Conference and Launching Ceremony of Go + Developer Foundation were grandly held in Shanghai. At the press conference, Xu Shiwei, CEO of Qiniuyun and inventor of Go + language, and Go + language contributors jointly released Go + version 1.0 and announced the development roadmap of Go +.

In order to maintain close communication with Go + developers and industry practitioners, and jointly promote and exchange the iterative development of Go +, the Go + core development team specially planned a series of live sharing of "Go + Open Class". The theme of the first issue was "Design and Implementation of Go + v1. x", and the sharing guest was Xu Shiwei, CEO of Seven Niuyun and inventor of Go +.

This article is based on the live broadcast content. Click on the end of the article to read the original text and view the full video playback.

Live preview: 

At 19:00 on November 13th (Saturday), the second issue of Go + Open Class officially started! Xu Shiwei, CEO of Qiniuyun and inventor of Go + language, will share "import process and module management of Go +". Click the button to make an appointment for live broadcast! Scan the code to enter the live exchange group and participate in the extraction of Go + official periphery ~

Today's sharing is the first lecture series on "Design and Implementation" after Go + 1.0, which mainly shares two parts:

- Go + Overall Architecture
- How to add functionality to Go +
	
The first part mainly shares how to realize the overall architecture of Go +; The second part is to talk with you about how to become a contributor to Go + and how to add new functions to Go +.
	
I. Go + Overall Architecture

The code of Go + is much simpler than that of Go. I have previously shared three ways of writing Hello World of Go + in the knowledge planet "Go + Open Class":

The simplest way is "imperative", which is more like code on the command line. 

According to my observation, when children understand programming statements, the threshold of "imperative" statements is the lowest, and even more acceptable than function calls. Therefore, although I struggled for a while, I decided to add "imperative" grammar to further lower the threshold and facilitate children to understand.

What I want to share today is what happens behind grammar. The following figure is the overall architecture diagram of Go +. Let's take Hello world as an example to see what happens during execution.

First, look at the overall architecture diagram on the left. The starting point is the source code of Go +, and the end point may be two: if you are running an executable file, the end point may be a software, and the other may be a Package.
	
In this process, we actually experienced several intermediate products, and the part in the graphic box above is the description of intermediate products, such as Go + Source, Go + Token, Go + AST (Abstract Syntax Tree), Go DOM Writer and so on.
	
Among them, the "DOM Writer" module in the cloud framework is not a concrete product. In simple words, you can "fill things" into it, and then you can generate Go AST, which can be converted into Go + code. Finally, it is packaged as a package or executable program through the command line program of Go +.
	
The overall process is roughly like this.

The arrows in the figure are marked with many blue characters, which represent the modules needed to really accomplish related things. On the right is the name of the key module. These names may seem familiar to everyone, and it may be easier for people who know the compilation principle to understand them. Let's explain them one by one.

1. Token & Scanner

The key concepts involved in the first step are Token and Scanner.

Token is a concept in compiling principle, which is similar to the concept of "word" in natural language. Source code is actually a byte stream, which is a string of text. Through a process similar to lexical analysis, the byte stream is converted into a stream of "words", which is called Token stream.
	
This process is often called "lexical analysis", or "lex" process, in compilation principles. In the code, we often call this module Scanner, which is actually the process of lexical analysis.
	
Usually, when we write our own code, we don't need to do lexical analysis by ourselves, but Parser is responsible for calling it. So this is actually a basic process, turning the text into words.

For example, "if" of an if statement is a word, and the integer "100" is a word. It is the concept of Token to cut into words with certain grammatical meanings like this.
	
Understanding the Token flow makes it very easy to understand the first step.

2. AST & Parser

The key concept involved in the second step is AST & Parser.

The process of this step is the process of grammatical analysis, and the upper level of morphology must be grammar, where the logic is similar to natural language.

The full name of AST is abstract syntax tree, which is the DOM of language. The result of our syntax analysis is a syntax tree, which corresponds to a structured article in natural language, that is, DOM tree. DOM model is a classic model in text processing.
	
By analogy, common documents that we often come into contact with, such as XML and json, also have their own DOM, which is essentially the same as languages having their own AST. Therefore, AST is actually a DOM, which meets the needs of our understanding of syntax structure.

Parser is actually grammar analysis, which is responsible for transforming token flow or word flow into grammar tree, that is, having an understanding of grammar.
	
These two processes, lexical parsing and syntax parsing, are very standard in compiling principles.

Let's take a look at how Parser is used.

Parser is used in the same way as Go in Go +:

fset: It is mainly used to record the relationship between file offset and line: col, and it is used to locate the prompt when syntax errors occur, such as which row and which column the errors appear in;

path: The most important parameter is the directory where the source code is located;

filter: filter file use, can pass nil. This parameter is not important and is used to filter out files in the directory that we don't want to deal with;

mode: Some flags that control the Parser process, which can be simply passed 0;

The above is the input related way, so you can see that the most important thing is the directory where the source code is located. Among the parameters related to the output, the first parameter is the most important:
	
pkgs: The resulting AST. Because there may be multiple pkgs in a directory, pkgs is a map;
first: The first error that occurs when Parser goes wrong.
	
3. cl & gox

The next stage is the narrow compilation process, which involves the cl & gox modules. We usually call the whole process compilation, which shows that this is the most critical process.

The essence of compilation is semantic analysis, which is responsible for transforming Go + AST into a call to gox DOM Writer. Here is a complexity brought about by Go compatibility-the derivation of types. For example, it is very difficult to deduce what kind of variable "variable a" belongs to in semantic analysis. This is also the biggest difference between Go + 1.0 and Go + 0.7-in the new version, we introduced the gox module, which is responsible for generating the abstract syntax tree (AST) of Go.
	
This part is extracted as a separate module because semantic analysis in Go has a key point, that is, "deriving variable types" mentioned above, which is a very complicated process, so we have built-in automatic derivation mechanism for variable types in gox, a DOM Writer component.

The Go + 0.7 version actually does automatic derivation of types in the compiler, which makes the compiler very complicated. After the new version optimizes this mechanism, it will be much easier to add functions to Go +, which makes the compiler's work very simple.
	
The compiler inputs the abstract syntax tree of Go + and outputs the DOM Writer of Go, and this component generated by Go is actually to generate the abstract syntax tree of Go. Therefore, we can see that the result of cl + gox is to transform Go + AST into Go AST, which essentially realizes the transformation of data format.
	
Let's make an analogy. Friends who know the invention history of C + + may know that the earliest compiler of C + + is called C-front, which is the front end of C. This version of the C + + compiler is to convert C + + into C, and then compile by C's compiler. At present, the working principle of Go + is very similar to C-front, which is very similar to a Go-front. Based on this pattern, we reuse the tool chain of Go language.
	
Let's look at how the compiler is used.

The cl procedure has three input parameters:

pkgPath: The target to compile, the import path of Go + pkg, that is, what the path name looks like when this package is referenced, which is a very important information in the compilation process;
pkg: AST of the target Go + pkg to compile;
conf: Configuration for compilation. The simplest thing is to pass nil, which enables the default configuration.
	
There are two parameters in the output:
	
p: Components of the generated gox DOM Writer. At this time, the returned component is not empty, but its interface has been called by cl to complete the format conversion and has been filled with data;
err: Returns all compilation errors if an error occurs during compilation. Note that unlike the previous parser procedure, parser returns only the first error because it is compatible with Go semantics, while the cl procedure returns all errors.
	
gox can be simply divided into two parts:

The first part is used to fill cl with data and realize format conversion;

The second part is used to generate Go AST/Source.

The function interface in the first part is very complex, so it is necessary to create a package, in which a function or type, variable or constant is created. After creating the function, it is necessary to write code through API. Therefore, there are a large number of interfaces in the format conversion part, which can basically constitute a complete DOM tree generation process, which is why we call gox module DOM Writer component, because most of its API is for generating DOM.

The second part is how to generate the abstract syntax tree of Go, or how to generate the source code of Go after filling the data.
	
The first function, ASTFile, is to generate Go AST, a very simple function call. Its first parameter is the gox DOM Writer instance. The second parameter is more interesting. You know that a directory often contains multiple packages, most commonly a normal code package and a package that generates test code. The second parameter is to determine what kind of syntax tree we want to generate, and then generate it.

The second function, WriteFile, is a simpler way to generate Go's source code file, with the first parameter, which is a file name, than just now. We use this file name to generate the source code of Go.
	
Once you have Go's source code or abstract syntax tree, you can call Go Tools to compile. For example, we can go run, go install, and of course you can go build.

Of course, at the beginning of Go + release, we said that we are dual-engine, and we support both static compilation and dynamic interpretation execution.

At present, Go + has implemented a dual-engine system, but the maturity of the two engines is temporarily different. Statically compiled engine engineering completion is very high, and only a limited number of Go functions are not supported for the time being.

But Go + 's interpreter is not yet very mature, and it is certainly much better than the interpreter engine we showed at last year's press conference. The current Go + interpreter adopts an open architecture, which supports a large number of Go parsers as the underlying engine. For example, use the Go parser made by the Go + team itself.

At present, the parser method of Go + is relatively simple, which converts the source code of Go + into the abstract syntax tree of Go +, then becomes the abstract syntax tree of Go, and finally executes it through the parser of Go.

At present, the completeness of Go interpreters in the market, including ourselves, is not particularly high, and we will further improve them in the future, especially when doing enhancement in the field of data science.

These are the core processes of the overall framework of Go +. Let's sum up the following picture:

It is a series of the whole content. As you can see, the whole core process is still very simple.

First, new a FlieSet, which is just the need for an error prompt. The real first step is the second line, which is the Parser procedure. We pass a "." that is, the current directory to Parser, and when Parser finishes, it gets pkgGops, which is a list of packages.

Let's assume that we are concerned with the main package, and we take out the Go + AST of the main package and pass it to the compiler for compilation.
	
So the third line is the compilation process. The compilation call is cl.NewPackage, and we pass in the abstract syntax tree of Go +, and finally get an instance of DOM Writer of Go. We call this instance for WriteFile, and we can generate the source code of Go. Finally, call "go run" and you can execute it.

The macro framework of Go + is roughly like this. Understanding the code on this page can basically connect the whole process in series.
	
2. How to add functionality to Go +

Next, we will analyze how to understand the architecture of Go + from the perspective of how to add functions to Go +. In this section, we will share which modules are usually involved and how to modify them.

1. Add syntax to Go +

Let's make a hypothesis: If we want to implement the so-called "trinocular operator" in C language, which modules do we need to modify? Usually we need to make a few judgments:
	
In the first step, we need to determine whether we have changed the abstract syntax tree of Go +. Of course, there will be functions that can be added without modifying the abstract syntax tree, but generally speaking, a new function often needs to modify the abstract syntax tree. For example, to implement trinocular operators, it is necessary to add a node of the abstract syntax tree to express the new language.
	
Modifying AST usually modifies two parts: abstract syntax tree and syntax parsing. The two are closely related because pasrser is the component responsible for generating the abstract syntax tree.
	
So the first thing we need to do when adding functionality is modify the ast module to add an abstract syntax tree node to it. Still taking the trinomial operator as an example, we need to add a node of the trinomial operator expression.
	
In the second step, after you have a new abstract syntax tree, you need to modify the parser module to escape the text of the trinocular operator into an abstract syntax tree.
	
The third step is to modify the cl module. The cl module is responsible for converting the new trinocular operator expression into a call to the DOM Writer component gox to generate an abstract syntax tree for Go. Because the trinocular operator is an expression, we are modifying the cl/expr.go file in the compiler. If a statement is added, it is usually the cl/stmt. go file that is modified.
	
Let's look at the specific contents to be modified in each module.
First, let's look at what we want to add to the ast module.
	
In the picture above, you can see "Cond? X: Y "is an abstract syntax tree with one-to-one text correspondence, which is very easy to understand. There are three very simple functions that the expression needs to implement in Go + AST, representing the starting point and ending point of the expression, and an empty function indicating that this is an expression node.

The second part is to modify the parser, that is, the parsing module. Syntax analysis is responsible for transforming token sequence into nodes of abstract syntax tree. This process is relatively personalized and does not have much common relationship. The location of link jump is the location of the code modified by trinocular operators. If you are interested, you can try to change it.

The third part is modifying the compiler. The first step in modifying the compiler is not to change it immediately, but to think in mind what the AST node of the added functionality would look like to convert to Go code. The blue annotation in the above figure is the trinocular operator, and the result of its final transformation I made a possible implementation, which is a call to a closure.
	
What are the difficulties in implementing this closure? It is the return value type "T" of closure. What type of T is cannot be judged by syntax alone, but by semantics. As we mentioned earlier, the type derivation is done through the DOM Writer component gox, and you can see how to derive the return value "T" with this component.
	
If we use the compiler, that is, the cl module, to derive the return value "T", the compiler code will be much more complicated, while the compiler code of Go + v1. x version is very concise, because we free the work of type derivation from the compiler, so its code is basically close to the feeling of parsing only.
	
The overall structure of the code in the above figure is very simple, and the only difficulty is how to generate "T".

The above figure is part of the compiler code finally edited. Among them, the code for compiling trinocular operator expressions is very short, basically only about ten lines.
	
Let's see how it is realized. First, in the first line, we define an automatic variable, that is, a variable whose type needs to be deduced automatically.
	
In the second step, we new a closure. The argument list for this closure is empty, so the first argument is nil; There is only one return value, so we create a Tuple to pass in the variable just defined; Because this closure is non-indefinite, the third parameter is false; BodyStart is the body of the function that starts the closure, starting to write the code.
	
The latter DOM is very simple, but you may find it strange: Why compile the expression before returning? Anyone who has studied compilation principles will know a concept-inverse Polish expression, which is the expression used by DOM, with parameters before operators.
	
Return has a parameter "1" to Return only one expression, and the last "end" to indicate the end of the closure. "call (0)" corresponds to the parentheses in the compiler modification section, indicating that the closure was called directly without passing parameters.
	
It's almost such a concept. On the whole, there is nothing particularly difficult to understand.
	
The reason why gox is very important is that when the compiler translates the abstract syntax tree of Go + and finally turns it into truly understandable code, it needs the help of gox DOM Writer.
	
If you want to contribute code to Go + and understand and implement the functional development of Go +, you need to understand the gox module very well, because basically the whole modification process is related to gox.
	
2. Write test cases

As you can see, it is very easy to add functions to Go + after having these functions and features. Of course, we attach great importance to engineering, and we must consider the test case after completing the function.

The abstract syntax tree, that is, the ast module part, basically does not need to be tested, because it only defines the data structure and has little code;

The parser module needs to be tested to see if it correctly transforms the Go + code into the desired abstract code tree.

The cl module needs to test whether the Go + code is correctly converted into the desired Go code.
	
Let's explain them one by one.

First, let's look at how to test the parser module. For parsing, in fact, we basically don't need to write test code, but add test cases to our prepared test framework. I think this is a point where Go + has done a good job in productivity optimization.
	
There is a _ under gop/parserThe testdata directory, which is full of test cases, and each test case is a directory. Let's still take the trinomial expression as an example. The directory can be named "ternary". There are two files in each test case directory-the first file has the suffix. gop, that is, the code you wrote, and you can take whatever file name you want, such as ternary.gop for trinocular operators. The other file name must be fixed as parser.expect, and its content is the result of the expected abstract syntax tree dump into text, which is used to verify whether our converted code is correct.
	
We have prepared three test cases for lambda expressions, so we can take a look at them and basically use them ourselves.
	
Of course, who executes the test cases? It's actually the function TestFromTestdata. But there is a problem here. This function tests a batch of test cases. When we find that there is something wrong with the code we added and want to debug it, we need to modify the TestFromTestdata function, and change its first line "sel: =" "" to "sel: =" ternary "" or other directory names of test cases we set, so that we can only execute the test cases we added.

Next is how to test the cl module, which is the semantic analysis part. This part is simpler than the parser part.
	
You basically just need to find compile_ in the compiler directoryTest. Go file, add a test function in it, such as TestTernaryExpr. The code of this function is also very simple. Call the gopClTest function and give you the code you want Go + and the code you want to generate Go.

Finally, we summarize the complete process of adding functions to Go +.
	
The first step is to fork Go + code first, and then clone your fork code repository to develop locally.

The first step of development suggests that you create a new function branch, and then modify the code and pass the test, then you can submit the code and submit the pull request to Go +.
	
The whole process of adding functionality to Go + is probably like this. 

Through the complete introduction of the above added functions, I would like to finally summarize the key contents:

Adding new features is usually modifying gop/ast, gop/parser and gop/cl modules
The most complex cl process is much easier when you add new functionality to Go +, thanks to gox DOM Writer's powerful type auto-derivation and Go AST generation capabilities

These two points are very important for the subsequent iterative development of Go +.

III. Exercises & Contact Information
	
Finally, leave some exercises for interested friends. You are also welcome to share the obstacles encountered during the practice, so that we can further simplify the process.

1. Basic exercises
Implement the trinocular operator (cond? expr1: expr2)

2. Advanced exercises

Special note: Realize the practice of trinocular operators we do not need to submit pull request, Go + temporarily no plans to add this function. But it can be used as a good exercise to practice and test.

We will also help you solve the problems encountered in the training process. Contact information is as follows:
	
1. Go + User Group (WeChat Group): You can ask questions directly in the group and @ me, I will answer them directly in the community;
2. Go + Open Class (Knowledge Planet): PPT and text content of this speech will be synchronized in Knowledge Planet. Welcome to ask questions and exchange on it.

This is a paragraph[Link](https://www.example.com)

This is a paragraph![Image](https://www.example.com/image.jpg)
 <nil>
--- PASS: TestTranslateMarkdown (0.70s)
PASS
ok  	github.com/goplus/community/translation	1.006s
</pre>
